cmd/libsnap-confine-private/snap.c

```diff
@@ -195,85 +309,82 @@ void sc_instance_key_validate(const char *instance_key, sc_error **errorp)
-	// expression engine.
-	const char *p = snap_name;
-	if (skip_one_char(&p, '-')) {
-		err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME,
-				    "snap name cannot start with a dash");
+
+	const char *pos = strchr(snap_component, '+');
+	if (pos == NULL) {
+		err =
+		    sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_COMPONENT,
+				  "snap component must contain a +");
 		goto out;
 	}
-	bool got_letter = false;
-	int n = 0, m;
-	for (; *p != '\0';) {
-		if ((m = skip_lowercase_letters(&p)) > 0) {
-			n += m;
-			got_letter = true;
-			continue;
-		}
-		if ((m = skip_digits(&p)) > 0) {
-			n += m;
-			continue;
-		}
-		if (skip_one_char(&p, '-') > 0) {
-			n++;
-			if (*p == '\0') {
-				err =
-				    sc_error_init(SC_SNAP_DOMAIN,
-						  SC_SNAP_INVALID_NAME,
-						  "snap name cannot end with a dash");
-				goto out;
-			}
-			if (skip_one_char(&p, '-') > 0) {
-				err =
-				    sc_error_init(SC_SNAP_DOMAIN,
-						  SC_SNAP_INVALID_NAME,
-						  "snap name cannot contain two consecutive dashes");
-				goto out;
-			}
-			continue;
-		}
-		err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME,
-				    "snap name must use lower case letters, digits or dashes");
+
+	size_t snap_name_len = pos - snap_component;
+	if (snap_name_len > SNAP_NAME_LEN) {
+		err =
+		    sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_COMPONENT,
+				  "snap name must be shorter than 40 characters");
 		goto out;
 	}
-	if (!got_letter) {
-		err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME,
-				    "snap name must contain at least one letter");
+
+	size_t component_name_len = strlen(pos + 1);
+	if (component_name_len > SNAP_NAME_LEN) {
+		err =
+		    sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_COMPONENT,
+				  "component name must be shorter than 40 characters");
 		goto out;
 	}
-	if (n < 2) {
-		err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME,
-				    "snap name must be longer than 1 character");
+
+	char snap_name[SNAP_NAME_LEN + 1] = { 0 };
+	strncpy(snap_name, snap_component, snap_name_len);
+
+	char component_name[SNAP_NAME_LEN + 1] = { 0 };
+	strncpy(component_name, pos + 1, component_name_len);
+
+	validate_as_snap_or_component_name(snap_name, SC_SNAP_INVALID_COMPONENT, "snap name in component", &err);
+	if (err != NULL) {
 		goto out;
 	}
-	if (n > SNAP_NAME_LEN) {
-		err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME,
-				    "snap name must be shorter than 40 characters");
+
+	validate_as_snap_or_component_name(component_name, SC_SNAP_INVALID_COMPONENT, "component name", &err);
+	if (err != NULL) {
 		goto out;
 	}
 
+	if (snap_instance != NULL) {
+		char snap_name_in_instance[SNAP_NAME_LEN + 1] = { 0 };
+		sc_snap_drop_instance_key(snap_instance, snap_name_in_instance,
+					  sizeof snap_name_in_instance);
+
+		if (strcmp(snap_name, snap_name_in_instance) != 0) {
+			err =
+			    sc_error_init(SC_SNAP_DOMAIN,
+					  SC_SNAP_INVALID_COMPONENT,
+					  "snap name in component must match snap name in instance");
```

Comment:
this bit isn't covered by tests

Reply:
Covered in d44445c0705422d0270cdfa59aa9a6e902e502a3.